File: game-server/src/main/java/com/ggumtle/ggumtle/server/packet/PacketHeader.java
package com.ggumtle.ggumtle.server.packet;

import java.nio.ByteBuffer;

/**
 * @param packetType 패킷 타입
 * @param dataLength 데이터 길이
 * @param timestamp  타임스탬프
 */
public record PacketHeader(
        short packetType,
        int dataLength,
        long timestamp
) {
    public static final int HEADER_SIZE = 14; // short + int + long = 2 + 4 + 8 = 14

    public static PacketHeader fromBytes(byte[] bytes) {
        if (bytes.length != HEADER_SIZE) {
            throw new IllegalArgumentException("Packet Header 길이가 맞지 않습니다.");
        }

        ByteBuffer buffer = ByteBuffer.wrap(bytes);

        // Java는 기본 Big Endian
        // 명시해두기 위해 작성해두었지만, 혹시 모를 성능을 대비해 주석 처리
        // buffer.order(ByteOrder.BIG_ENDIAN);

        short packetType = buffer.getShort();
        int dataLength = buffer.getInt();
        long timestamp = buffer.getLong();

        return new PacketHeader(packetType, dataLength, timestamp);
    }

    public byte[] toBytes() {
        ByteBuffer buffer = ByteBuffer.allocate(HEADER_SIZE);

        // Java는 기본 Big Endian
        // 명시해두기 위해 작성해두었지만, 혹시 모를 성능을 대비해 주석 처리
        // buffer.order(ByteOrder.BIG_ENDIAN);

        buffer.putShort(packetType);
        buffer.putInt(dataLength);
        buffer.putLong(timestamp);

        return buffer.array();
    }
}

File: game-load-test/src/main/kotlin/com/ggumtle/loadtest/protocol/PacketHeader.kt
package com.ggumtle.loadtest.protocol

import io.netty.buffer.ByteBuf
import java.nio.ByteBuffer

/**
 * Packet header structure (14 bytes, Big Endian)
 *
 * [0-1]   short - Packet Type
 * [2-5]   int   - Data Length
 * [6-13]  long  - Timestamp
 */
data class PacketHeader(
    val packetType: Short,
    val dataLength: Int,
    val timestamp: Long
) {
    companion object {
        const val HEADER_SIZE = 14 // short(2) + int(4) + long(8)

        fun fromByteBuf(buf: ByteBuf): PacketHeader {
            return PacketHeader(
                packetType = buf.readShort(),
                dataLength = buf.readInt(),
                timestamp = buf.readLong()
            )
        }

        fun fromBytes(bytes: ByteArray): PacketHeader {
            require(bytes.size == HEADER_SIZE) { "Invalid header size: ${bytes.size}" }
            val buffer = ByteBuffer.wrap(bytes)
            return PacketHeader(
                packetType = buffer.short,
                dataLength = buffer.int,
                timestamp = buffer.long
            )
        }
    }

    fun writeTo(buf: ByteBuf) {
        buf.writeShort(packetType.toInt())
        buf.writeInt(dataLength)
        buf.writeLong(timestamp)
    }

    fun toBytes(): ByteArray {
        return ByteBuffer.allocate(HEADER_SIZE)
            .putShort(packetType)
            .putInt(dataLength)
            .putLong(timestamp)
            .array()
    }
}
